mlir/include/mlir/Interfaces/LoopLikeInterface.td

```diff
@@ -220,6 +220,31 @@ def LoopLikeOpInterface : OpInterface<"LoopLikeOpInterface"> {
       /*defaultImplementation=*/[{
         return ::mlir::failure();
       }]
+    >,
+    InterfaceMethod<[{
+        Add a zero-trip-check around the loop to check if the loop body is ever
```

Comment:
Instead of creating a "check" (it is not really clear what the "check" is; it could be an `scf.if`, it could be a new basic block with a conditional branch, etc.), this method could also return just the condition. (Which evaluates to `true` if the loop has at least one iteration.) That would keep the interface a bit simpler. What do you think?

Reply:
I think the trick part of adding zero-trip-check is to do loop rotation when the loop condition has side-effects. So if we simply return the loop condition, callers will need to implement loop rotation by themselves for each loop op, which can be complicated.

Reply:
This makes sense, I did not think of side effects.

The lambda sounds like a good idea, but I'm not sure what exactly it should look like. (Should it return a `Block *` into which the loop is moved?) We can generalize later if needed.

I think it would also work for `scf.for`. The check condition is `lb < ub`.

Reply:
I think whether the loop rotation is needed (or possible) is the implementation details of each type of loop. `scf.while` needs it because the before block is in the loop and can contain ops with side-effects. `scf.for` is simpler as the `lb < ub` should have no side-effect and lightweight, which can be run twice (for `scf.if` and the for loop)